packages/react-sdk/src/components/image.tsx

```diff
@@ -1,24 +1,187 @@
 import React, { forwardRef, type ElementRef, type ComponentProps } from "react";
+import { warnOnce } from "../utils/warn-once";
 
 const defaultTag = "img";
 
-type ImageProps = ComponentProps<typeof defaultTag>;
+type ImageProps = ComponentProps<typeof defaultTag> & { quality?: number };
+
+/**
+ * max(...imageSizes) must be less then min(...deviceSizes)
+ * idea is to support images with vw size less than 100vw
+ * https://github.com/vercel/next.js/blob/canary/packages/next/shared/lib/image-config.ts
+ **/
+const imageSizes = [16, 32, 48, 64, 96, 128, 256, 384];
+const deviceSizes = [640, 750, 828, 1080, 1200, 1920, 2048, 3840];
+
+const allSizes = [...imageSizes, ...deviceSizes];
+
+/**
+ * https://developers.cloudflare.com/images/image-resizing/url-format/
+ **/
+const cloudflareImageLoader = ({
+  width,
+
+  src,
+  quality,
+}: {
+  width: number;
+  src: string;
+  quality?: number;
+}) => {
+  let resizeOrigin = "";
+
+  if (process.env.NODE_ENV !== "production") {
+    if (!allSizes.includes(width)) {
+      warnOnce("Width must be only from allowed values");
+    }
+
+    // No image transformation in development
+    resizeOrigin = "https://webstudio.is";
+  }
+
+  const params = [
+    `width=${width}`,
+    quality != null && `quality=${quality}`,
+    // TODO: Must be checked how it works with caching (like avif image is cached and then read by the browser without avif support)
+    "format=auto",
+  ].filter(Boolean);
+
+  return `${resizeOrigin}/cdn-cgi/image/${params.join(",")}/${src}`;
+};
+
+/**
+ * https://github.com/vercel/next.js/blob/canary/packages/next/client/image.tsx
+ **/
+const getWidths = (
+  width: number | undefined,
+  sizes: string | undefined
+): { widths: number[]; kind: "w" | "x" } => {
+  if (sizes) {
+    // Find all the "vw" percent sizes used in the sizes prop
+    const viewportWidthRe = /(^|\s)(1?\d?\d)vw/g;
+    const percentSizes = [];
+    for (let match; (match = viewportWidthRe.exec(sizes)); match) {
+      percentSizes.push(parseInt(match[2]));
```

Comment:
wow, if they made simple mistakes like that, maybe we shouldn't copy their code at all, also we need to check the license, when we copy even if its MIT, we need to do attributions in proper places
I would rather rewrite it

Reply:
radix added